OpenRA.Mods.Common/Traits/Render/WithSwitchableOverlay.cs

```diff
@@ -0,0 +1,192 @@
+
+			base.RulesetLoaded(rules, ai);
+		}
+
+		public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer init, string image, int facings, PaletteReference p)
+		{
+			if (!EnabledByDefault)
+				yield break;
+
+			if (Palette != null)
+				p = init.WorldRenderer.Palette(IsPlayerPalette ? Palette + init.Get<OwnerInit>().InternalName : Palette);
+
+			Func<WAngle> facing;
+			var dynamicfacingInit = init.GetOrDefault<DynamicFacingInit>();
+			if (dynamicfacingInit != null)
+				facing = dynamicfacingInit.Value;
+			else
+			{
+				var f = init.GetValue<FacingInit, WAngle>(WAngle.Zero);
+				facing = () => f;
+			}
+
+			var anim = new Animation(init.World, Image ?? image, facing)
+			{
+				IsDecoration = IsDecoration
+			};
+
+			if (!string.IsNullOrEmpty(EnabledSequence))
+				anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), EnabledSequence));
+			else if (!string.IsNullOrEmpty(DisabledSequence))
+				anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), DisabledSequence));
+			else if (!string.IsNullOrEmpty(SwitchingSequence))
+				anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), SwitchingSequence));
+
+			var body = init.Actor.TraitInfo<BodyOrientationInfo>();
+			WRot Orientation() => body.QuantizeOrientation(WRot.FromYaw(facing()), facings);
+			WVec Offset() => body.LocalToWorld(this.Offset.Rotate(Orientation()));
+			int ZOffset()
+			{
+				var tmpOffset = Offset();
+				return tmpOffset.Y + tmpOffset.Z + 1;
+			}
+
+			yield return new SpriteActorPreview(anim, Offset, ZOffset, p);
+		}
+	}
+
+	public class WithSwitchableOverlay : PausableConditionalTrait<WithSwitchableOverlayInfo>, ITick, INotifyDamageStateChanged
+	{
+		readonly Animation overlay;
+		int switchingLevel;
+		int chargeSpeed;
+		bool hasSwitched;
+
+		public WithSwitchableOverlay(Actor self, WithSwitchableOverlayInfo info)
+			: base(info)
+		{
+			switchingLevel = info.SwitchingLevelOnSpawn;
+
+			var rs = self.Trait<RenderSprites>();
+			var body = self.Trait<BodyOrientation>();
+			var facing = self.TraitOrDefault<IFacing>();
+
+			var image = info.Image ?? rs.GetImage(self);
+			overlay = new Animation(self.World, image, facing == null ? () => WAngle.Zero : (body == null ? () => facing.Facing : () => body.QuantizeFacing(facing.Facing)), () => false)
+			{
+				IsDecoration = info.IsDecoration
+			};
+
+			if (!string.IsNullOrEmpty(Info.EnabledSequence))
+				overlay.PlayRepeating(RenderSprites.NormalizeSequence(overlay, self.GetDamageState(), Info.EnabledSequence));
+			else if (!string.IsNullOrEmpty(Info.DisabledSequence))
+				overlay.PlayRepeating(RenderSprites.NormalizeSequence(overlay, self.GetDamageState(), Info.DisabledSequence));
+			else if (!string.IsNullOrEmpty(Info.SwitchingSequence))
+				overlay.PlayRepeating(RenderSprites.NormalizeSequence(overlay, self.GetDamageState(), Info.SwitchingSequence));
```

Comment:
I'm just wondering. Does this pattern exist anywhere else? It  feels a bit wrong for it to not be handling conditions. Perhaps it should be handled in INotifyCreated?

Reply:
Nah, it just because if there is no sequence being played, `AnimationWithOffset` will crash:
```
var anim = new AnimationWithOffset(overlay, ...);
```

The lines here try to find out a valid sequence, or any valid sequence to fill the blank since you must allow a null value for all sequences.

Reply:
That is the same even you move it to "INotifyCreated", and it is not related to conditions.